core/src/note.rs

```diff
@@ -454,30 +410,164 @@ impl Serializable<SIZE> for Note {
             bytes[0].try_into().map_err(|_| BytesError::InvalidData)?;
 
         let mut buf = &bytes[1..];
+
         let value_commitment = JubJubAffine::from_reader(&mut buf)?;
+
         let stealth_address = StealthAddress::from_reader(&mut buf)?;
+
         let pos = u64::from_reader(&mut buf)?;
 
         let mut value_enc = [0u8; VALUE_ENC_SIZE];
         value_enc.copy_from_slice(&buf[..VALUE_ENC_SIZE]);
-
         buf = &buf[VALUE_ENC_SIZE..];
 
-        let sender_enc_A_0 = JubJubAffine::from_reader(&mut buf)?;
-        let sender_enc_A_1 = JubJubAffine::from_reader(&mut buf)?;
-        let sender_enc_B_0 = JubJubAffine::from_reader(&mut buf)?;
-        let sender_enc_B_1 = JubJubAffine::from_reader(&mut buf)?;
+        let sender = Sender::from_reader(&mut buf)?;
 
         Ok(Note {
             note_type,
             value_commitment,
             stealth_address,
             pos,
             value_enc,
-            sender_enc: [
-                (sender_enc_A_0, sender_enc_A_1),
-                (sender_enc_B_0, sender_enc_B_1),
-            ],
+            sender,
         })
     }
 }
+
+/// The sender of the `Note`.
+/// This can be either the encrypted sender's [`PublicKey`], if the [`Note`] was
+/// created as an output note of a phoenix-transaction, or some contract-data if
+/// the [`Note`] was created in another way, e.g. by withdrawing from a
+/// contract.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "rkyv-impl",
+    derive(Archive, Serialize, Deserialize),
+    archive_attr(derive(bytecheck::CheckBytes))
+)]
+pub enum Sender {
+    /// The sender's [`PublicKey`], encrypted using the note_pk of the
+    /// stealth-address.
+    Encryption([(JubJubAffine, JubJubAffine); 2]),
+    /// Information to identify the origin of a `Note`, if it wasn't created as
+    /// a phoenix-transaction output-note.
+    ContractInfo([u8; 4 * JubJubAffine::SIZE]),
```

Comment:
I'm not sure I understand the usage of `JubJubAffine::SIZE` here. What is the intent?

Reply:
The contract info is supposed to be the same size as four `JubJubAffine` would take otherwise.

Reply:
This way the sender is always constant size

Reply:
> This way the sender is always constant size and we can implement `Serializable` on it.

Ah got it!